sdk/internal/testframework/recording.go

```diff
@@ -0,0 +1,371 @@
+	envMode := GetOptionalEnv(ModeEnvironmentVariableName, string(mode))
+	mode = RecordMode(envMode)
+
+	// initialize the Recording
+	recording := &Recording{
+		SessionName:              recPath,
+		RecordingFile:            recPath + ".yaml",
+		VariablesFile:            varPath,
+		Mode:                     mode,
+		variables:                make(map[string]string),
+		previousSessionVariables: make(map[string]string),
+		recorder:                 rec,
+		c:                        c,
+	}
+
+	// Try loading the recording if it already exists to hydrate the variables
+	err = recording.initVariables()
+	if err != nil {
+		return nil, err
+	}
+
+	// set the recorder Matcher
+	rec.SetMatcher(recording.matchRequest)
+
+	// wire up the sanitizer
+	DefaultSanitizer(rec)
+
+	return recording, err
+}
+
+// Gets a recorded variable. If the variable is not found we panic
+// variableType is optional and defaults to Default
+func (r *Recording) GetRecordedVariable(name string, variableType ...VariableType) string {
+	result, ok := r.previousSessionVariables[name]
+	if !ok || r.Mode == Live {
+		result = GetRequiredEnv(name)
+		r.variables[name] = applyVariableOptions(&result, variableType...)
+	}
+	return result
+}
+
+// Gets a recorded variable with a fallback default value
+// variableType is optional and defaults to Default
+func (r *Recording) GetOptionalRecordedVariable(name string, defaultValue string, variableType ...VariableType) string {
+	result, ok := r.previousSessionVariables[name]
+	if !ok || r.Mode == Live {
+		result = GetOptionalEnv(name, defaultValue)
+		r.variables[name] = applyVariableOptions(&result, variableType...)
+	}
+	return result
+}
+
+// To satisfy the azcore.Transport interface
+func (r *Recording) Do(req *http.Request) (*http.Response, error) {
+	resp, err := r.recorder.RoundTrip(req)
+	if err == cassette.ErrInteractionNotFound {
+		error := missingRequestError(req)
+		r.c.Fail(error)
+	}
+	return resp, err
+}
+
+// Stops the recording and saves them, including any captured variables, to disk
+func (r *Recording) Stop() error {
+
+	r.recorder.Stop()
+	if r.Mode == Live {
+		return nil
+	}
+
+	if len(r.variables) > 0 {
+		// Save the variables after merging them
+
+		//TODO: Merge values from previousVariables that are not in variables to variables
+
+		// Marshal to YAML and save variables
+		data, err := yaml.Marshal(r.variables)
+		if err != nil {
+			return err
+		}
+
+		f, err := r.createVariablesFileIfNotExists()
+		if err != nil {
+			return err
+		}
+
+		defer f.Close()
+
+		// http://www.yaml.org/spec/1.2/spec.html#id2760395
+		_, err = f.Write([]byte("---\n"))
+		if err != nil {
+			return err
+		}
+
+		_, err = f.Write(data)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Gets an environment variable by name and panics if it is not found
+func GetRequiredEnv(name string) string {
+	env, ok := os.LookupEnv(name)
+	if ok {
+		return env
+	} else {
+		panic(envNotExistsError(name))
```

Comment:
Panicking is a bit controversial. I know it's a feature of the language, but most people prefer not to panic. Although this will be an internal module, other Go SDK devs may use this so I would suggest following the other pattern and returning an error. So the return signature would look like `(string, error)`. 

You can also ask Joel and Jeff for their opinions on this.

Reply:
I think I should just call the testContext failed here instead. That would just fail the test.